cache.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2023 Adam Tauber, Andrzej Lichnerowicz
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package colly
+
+import (
+	"crypto/sha1"
+	"encoding/gob"
+	"encoding/hex"
+	"errors"
+	"net/http"
+	"os"
+	"path"
+)
+
+// Cache is an interface which handles caching Collector's responses
+// The default Cache of the Collector is the NullCache.
+// FileSystemCache keeps compatibility with non-pluggable caching in legacy
+// Collector. For this reason, one can set cache folder via `CACHE_DIR`
+// environment variable, or by passing CacheDir to NewCollector.
+// Collector's caching backend can be changed too by calling new method
+// Collector.SetCache
+type Cache interface {
+	// Init initializes the caching backend
+	Init() error
+	// Get retrieves a previously cached response for the given request
+	Get(request *http.Request) (*Response, error)
+	// Put stores a response given request in cache
+	Put(request *http.Request, response *Response) error
+	// Close finalizes caching backend
+	Close() error
```

Comment:
What's the expected lifecycle of the `Cache` implementation? Is it tightly coupled to the `Collector`'s one, or it can be shared between several collectors, and something else must close it?

Reply:
That's a very good question. The backwards-compatible `FileSystemCache` is tightly coupled with `Collector`, but the way to use other caches would be via `Collector.SetCache` only. Unfortunately in current form, where it's `.SetCache` who's calling `cache.Init` it kind of defeats the purpose. I will remove `.Init` from the `.SetCache` and document that this method should pass an initialised object, and a caller is responsible for closing. For the backwards-compatibility scenario, it should not matter too much. `Collector` does not have a `.Close` method, and `FileSystemCache` did not do any *closing* either.

Reply:
I thought about this a bit.

I believe if cache is set through the environment variable, collector's `Close` should call the cache's `Close` as well. If cache is initialized through `SetCache`, then it should not be closed. Moreover, if there's a cache previously initialized through the environment variable when `SetCache` is called, it should be closed before replacing it with the new one.

Do you agree?